{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1650f3ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format='retina'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "44420819",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import seaborn as sns\n",
    "import scrublet as scr\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f1b80e5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8927db44",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_colors = plt.cm.colors.ListedColormap(['red', 'darkred', 'blue','darkblue', 'orange', 'darkorange', 'violet', 'darkviolet',])\n",
    "\n",
    "samples = [\n",
    "    \"GSM6321073_PFC_4wk_1_matrix.h5\",\n",
    "    \"GSM6321077_PFC_90wk_1_matrix.h5\",\n",
    "]\n",
    "\n",
    "mouse_id = {\n",
    "    0 : 1,\n",
    "    1 : 1\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3d376ecf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM6321073 PFC 4wk 1\n",
      "reading /Users/cmdb/qb25project/mouse-brain-RNAseq/GSE207848_RAW/GSM6321073_PFC_4wk_1_matrix.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qb25project/lib/python3.13/site-packages/anndata/_core/anndata.py:1793: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " (0:00:01)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qb25project/lib/python3.13/site-packages/anndata/_core/anndata.py:1793: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM6321077 PFC 90wk 1\n",
      "reading /Users/cmdb/qb25project/mouse-brain-RNAseq/GSE207848_RAW/GSM6321077_PFC_90wk_1_matrix.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qb25project/lib/python3.13/site-packages/anndata/_core/anndata.py:1793: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " (0:00:01)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qb25project/lib/python3.13/site-packages/anndata/_core/anndata.py:1793: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "all_adata = []\n",
    "i = 0\n",
    "for s in samples:\n",
    "    label, area, age, idx, _ = s.split(\"_\")\n",
    "    print(label, area, age, idx)\n",
    "    curr_adata = sc.read_10x_h5(f\"/Users/cmdb/qb25project/mouse-brain-RNAseq/GSE207848_RAW/{s}\")\n",
    "    curr_adata.var_names_make_unique()\n",
    "    curr_adata.obs['area'] = area\n",
    "    curr_adata.obs['age'] = age\n",
    "    curr_adata.obs['idx'] = i\n",
    "    i += 1\n",
    "    curr_adata.var['mt'] = curr_adata.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    sc.pp.calculate_qc_metrics(curr_adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "    all_adata.append(curr_adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c6f848e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total cells: 28818\n"
     ]
    }
   ],
   "source": [
    "total_cells = np.sum([a.n_obs for a in all_adata])\n",
    "print('total cells:', total_cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35e9b6de",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qb25project/lib/python3.13/site-packages/anndata/_core/anndata.py:1791: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata = ad.concat(all_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "475bd7a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 28818 × 32285\n",
       "    obs: 'area', 'age', 'idx', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[adata.obs.area=='PFC']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "454561c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "756692c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 1766 cells that have less than 1000 genes expressed\n",
      "filtered out 24 cells that have more than 100000 counts\n",
      "filtered out 6412 genes that are detected in less than 3 cells\n",
      "filtered out 1369 cells that have less than 2500 counts\n"
     ]
    }
   ],
   "source": [
    "sc.pp.filter_cells(adata, min_genes=1000)\n",
    "sc.pp.filter_cells(adata, max_counts=100000)\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "sc.pp.filter_cells(adata, min_counts=2500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "64600e5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['mouse_id'] = [mouse_id[i] for i in adata.obs.idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b395fb02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 9.2%\n",
      "Estimated detectable doublet fraction = 66.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 9.0%\n",
      "\tEstimated  = 13.9%\n",
      "Elapsed time: 22.9 seconds\n"
     ]
    }
   ],
   "source": [
    "# run scrublet on adata to identify doublets\n",
    "scrub = scr.Scrublet(adata.X, expected_doublet_rate=0.09)\n",
    "doublet_scores, predicted_doublets = scrub.scrub_doublets(min_gene_variability_pctl=85, \n",
    "                                                          n_prin_comps=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "924f02c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.09174168907595776)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(predicted_doublets)/len(doublet_scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "37cac232",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~predicted_doublets,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3fb2a6d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qb25project/lib/python3.13/site-packages/anndata/_core/anndata.py:1175: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/opt/anaconda3/envs/qb25project/lib/python3.13/site-packages/anndata/_core/anndata.py:1175: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    }
   ],
   "source": [
    "adata.write(\"adata_combined_nodoublet.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ba66a8c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3332.0\n"
     ]
    }
   ],
   "source": [
    "print(np.median(adata.obs.n_genes_by_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "65ea09b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9699.0\n"
     ]
    }
   ],
   "source": [
    "print(np.median(adata.obs.total_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "220e6827",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata = adata[adata.obs.n_genes_by_counts < 3000, :]\n",
    "#adata = adata[adata.obs.pct_counts_mt < 5, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "09bb8ce4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "extracting highly variable genes\n",
      "    finished (0:00:00)\n",
      "--> added\n",
      "    'highly_variable', boolean vector (adata.var)\n",
      "    'means', float vector (adata.var)\n",
      "    'dispersions', float vector (adata.var)\n",
      "    'dispersions_norm', float vector (adata.var)\n"
     ]
    }
   ],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3cae6d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"adata_combined_nodoublet_normalized.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "819b65b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/m9/f9b1skd55v35_qp4hcflk2qh0000gp/T/ipykernel_21993/1433449232.py:22: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  f.write(adata.obs_names[i] + \"\\t\" + adata.obs[\"age\"][i] + \"\\n\")\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "adata = ad.read_h5ad('adata_combined_nodoublet.h5ad')\n",
    "\n",
    "#print(adata.obs.tail(n= 50))  # This shows the last few rows of cell metadata (annotations)\n",
    "\n",
    "adata.obs.dtypes\n",
    "\n",
    "# for i in adata.X:\n",
    "#     print(i)\n",
    "\n",
    "# print(adata.obs[\"age\"])\n",
    "# for i in adata.obs[\"age\"]:\n",
    "#     print(i)\n",
    "\n",
    "#print(adata.obs[\"n_genes\"])\n",
    "#print(adata.obs[\"n_genes_by_counts\"])\n",
    "\n",
    "with open ('age.tsv', \"w\") as f:\n",
    "    f.write(\"idx\" + \"\\t\" + \"age\" + \"\\n\")\n",
    "    for i in range(len(adata.obs['age'])):\n",
    "        #print(adata.obs_names[i], adata.obs[\"age\"][i])\n",
    "        f.write(adata.obs_names[i] + \"\\t\" + adata.obs[\"age\"][i] + \"\\n\")\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qb25project",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
