{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ccebff2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/anndata/_core/anndata.py:1798: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/anndata/_core/anndata.py:1798: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/anndata/_core/anndata.py:1798: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/anndata/_core/anndata.py:1798: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/anndata/_core/anndata.py:1796: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 9.2%\n",
      "Estimated detectable doublet fraction = 66.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 9.0%\n",
      "\tEstimated  = 13.9%\n",
      "Elapsed time: 26.9 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/scanpy/preprocessing/_normalization.py:269: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/scanpy/preprocessing/_pca/__init__.py:245: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.\n",
      "  Version(ad.__version__) < Version(\"0.9\")\n",
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/scanpy/neighbors/__init__.py:430: FutureWarning: Use obsm (e.g. `k in adata.obsm` or `adata.obsm.keys() | {'u'}`) instead of AnnData.obsm_keys, AnnData.obsm_keys is deprecated and will be removed in the future.\n",
      "  if \"X_diffmap\" in adata.obsm_keys():\n",
      "/var/folders/m9/f9b1skd55v35_qp4hcflk2qh0000gp/T/ipykernel_1344/2632080950.py:100: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  f.write(adata.obs_names[i] + \"\\t\" + adata.obs[\"age\"][i] + \"\\n\")\n",
      "/var/folders/m9/f9b1skd55v35_qp4hcflk2qh0000gp/T/ipykernel_1344/2632080950.py:108: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  f.write(adata.obs_names[i] + \"\\t\" + adata.obs[\"predicted_cc_type\"][i] + \"\\n\")\n",
      "/opt/anaconda3/envs/qbproject/lib/python3.13/site-packages/anndata/_io/utils.py:243: FutureWarning: Forward slashes will be disallowed in h5 stores in the next minor release\n",
      "  return func(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import scrublet as scr\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.sparse import csr_matrix\n",
    "\n",
    "# Input samples\n",
    "samples = [\n",
    "   \"GSM6321073_PFC_4wk_1_matrix.h5\",\n",
    "   \"GSM6321077_PFC_90wk_1_matrix.h5\",\n",
    "]\n",
    "\n",
    "# Mouse mapping\n",
    "mouse_id = {0: 1, 1: 1}\n",
    "\n",
    "# Load and process samples\n",
    "all_adata = []\n",
    "i = 0\n",
    "for s in samples:\n",
    "    label, area, age, idx, _ = s.split(\"_\")\n",
    "    curr_adata = sc.read_10x_h5(f\"/Users/cmdb/qb25project/mouse-brain-RNAseq/GSE207848_RAW/{s}\")\n",
    "    curr_adata.var_names_make_unique()\n",
    "\n",
    "    curr_adata.obs[\"area\"] = area\n",
    "    curr_adata.obs[\"age\"] = age\n",
    "    curr_adata.obs[\"idx\"] = i\n",
    "    i += 1\n",
    "\n",
    "    curr_adata.var[\"mt\"] = curr_adata.var_names.str.startswith(\"mt-\")\n",
    "    sc.pp.calculate_qc_metrics(curr_adata, qc_vars=[\"mt\"], inplace=True)\n",
    "\n",
    "    all_adata.append(curr_adata)\n",
    "\n",
    "# Concatenate\n",
    "adata = ad.concat(all_adata)\n",
    "adata.obs_names_make_unique()\n",
    "\n",
    "# Filtering\n",
    "sc.pp.filter_cells(adata, min_genes=1000)\n",
    "sc.pp.filter_cells(adata, max_counts=100000)\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "sc.pp.filter_cells(adata, min_counts=2500)\n",
    "\n",
    "# Mouse ID\n",
    "adata.obs[\"mouse_id\"] = [mouse_id[i] for i in adata.obs.idx]\n",
    "\n",
    "# Scrublet doublet removal\n",
    "scrub = scr.Scrublet(adata.X, expected_doublet_rate=0.09)\n",
    "doublet_scores, predicted_doublets = scrub.scrub_doublets(min_gene_variability_pctl=85, n_prin_comps=30)\n",
    "adata = adata[~predicted_doublets, :]\n",
    "\n",
    "# Normalize + log\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# HVGs, PCA, neighbors, Leiden, UMAP\n",
    "sc.pp.highly_variable_genes(adata)\n",
    "sc.tl.pca(adata)\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=20)\n",
    "sc.tl.leiden(adata, resolution=0.6)\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# Cell type scoring pipeline\n",
    "\n",
    "marker_genes = {\n",
    "   \"Excitatory Neurons\": [\"SLC17A7\", \"CAMK2A\", \"GRIN2A\"],\n",
    "   \"Macrophages\": [\"CD68\", \"ITGAX\", \"CSF1R\"],\n",
    "   \"Oligodendrocytes\": [\"OLIG1\", \"OLIG2\", \"MBP\"],\n",
    "   \"Microglia\": [\"CX3CR1\", \"ITGAX\", \"P2RY12\"],\n",
    "   \"OPC\": [\"PDGFRA\", \"SOX10\", \"OLIG2\"],\n",
    "   \"Peri/VLMC\": [\"PDGFRB\", \"COL1A1\", \"ACTA2\"],\n",
    "   \"Astrocytes\": [\"GFAP\", \"ALDH1L1\", \"SLC1A3\"],\n",
    "   \"MSN\": [\"DRD1\", \"PVALB\", \"SLC6A3\"],\n",
    "   \"Inhibitory Neurons\": [\"GAD1\", \"GAD2\", \"SLC32A1\"]\n",
    "}\n",
    "\n",
    "# Upper-case gene names\n",
    "adata.var_names = [g.upper() for g in adata.var_names]\n",
    "\n",
    "# Score gene signatures\n",
    "for cell_type, genes in marker_genes.items():\n",
    "    valid_genes = [g for g in genes if g in adata.var_names]\n",
    "    if len(valid_genes) > 0:\n",
    "        sc.tl.score_genes(adata, valid_genes, score_name=f\"{cell_type}_score\")\n",
    "\n",
    "# Assign predicted cell types\n",
    "score_cols = [col for col in adata.obs.columns if col.endswith(\"_score\")]\n",
    "adata.obs[\"predicted_cc_type\"] = adata.obs[score_cols].idxmax(axis=1).str.replace(\"_score\", \"\")\n",
    "\n",
    "# Create age.tsv \n",
    "with open(\"/Users/cmdb/qb25project/mouse-brain-RNAseq/age.tsv\", \"w\") as f:\n",
    "    f.write(\"idx\\tage\\n\")\n",
    "    for i in range(len(adata.obs[\"age\"])):\n",
    "        f.write(adata.obs_names[i] + \"\\t\" + adata.obs[\"age\"][i] + \"\\n\")\n",
    "\n",
    "# Create cell_types.tsv\n",
    "with open(\"/Users/cmdb/qb25project/mouse-brain-RNAseq/cell_types.tsv\", \"w\") as f:\n",
    "    f.write(\"idx\\tcell_type\\n\")\n",
    "    for i in range(len(adata.obs[\"predicted_cc_type\"])):\n",
    "        f.write(adata.obs_names[i] + \"\\t\" + adata.obs[\"predicted_cc_type\"][i] + \"\\n\")\n",
    "\n",
    "# Make .h5ad file for R \n",
    "# Ensure sparse matrix\n",
    "adata.X = csr_matrix(adata.X)\n",
    "\n",
    "# Make obs/var strings\n",
    "for col in adata.obs.columns:\n",
    "    adata.obs[col] = adata.obs[col].astype(str)\n",
    "\n",
    "for col in adata.var.columns:\n",
    "    if adata.var[col].dtype == \"object\":\n",
    "        adata.var[col] = adata.var[col].astype(str)\n",
    "\n",
    "# Remove everything that breaks Seurat\n",
    "adata.uns = {}\n",
    "\n",
    "# Save final file\n",
    "adata.write(\"/Users/cmdb/qb25project/mouse-brain-RNAseq/adata_for_seurat.h5ad\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qbproject",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
